packages/contracts-bedrock/tasks/genesis-l2.ts

```diff
@@ -46,6 +58,125 @@ const getStorageLayout = async (
+  const sources = buildInfo.output.sources
+  const contracts = buildInfo.output.contracts
+
+  const compilerOutputContracts: CompilerOutputContract[] = []
+  for (const [contractName, contract] of Object.entries(contracts)) {
+    if (path.basename(contractName, '.sol') === name) {
+      compilerOutputContracts.push(contract[name])
+    }
+  }
+  if (compilerOutputContracts.length === 0) {
+    throw new Error(`Cannot find compiler output contract for ${name}`)
+  }
+  if (compilerOutputContracts.length !== 1) {
+    console.log(`Unexpected number of contracts for ${name}`)
+  }
+  const outputContract = compilerOutputContracts[0]
+
+  const compilerOutputSources: CompilerOutputSource[] = []
+  for (const [contractName, source] of Object.entries(sources)) {
+    if (path.basename(contractName, '.sol') === name) {
+      compilerOutputSources.push(source as CompilerOutputSource)
+    }
+  }
+  if (compilerOutputSources.length === 0) {
+    throw new Error(`Cannot find compiler output source for ${name}`)
+  }
+  if (compilerOutputSources.length !== 1) {
+    console.log(`Unexpected number of sources for ${name}`)
+  }
+  const outputSource = compilerOutputSources[0]
+
+  return { outputContract, outputSource }
+}
+
+const replaceImmutables = async (
+  hre: HardhatRuntimeEnvironment,
+  name: string,
+  immutables: object
+): Promise<string> => {
+  const artifact = await hre.artifacts.readArtifact(name)
+  const buildInfo = await hre.artifacts.getBuildInfo(name)
+
+  const { outputContract, outputSource } = findContractAndSource(
+    name,
+    buildInfo
+  )
+
+  // Get the immutable references. They look like this:
+  // { ast-id: [ {start, length} ] }
+  const immutableReferences =
+    outputContract.evm.deployedBytecode.immutableReferences
+
+  const names = {}
+  // Recursively fine all of the immutables by traversing the solc output ast
+  const findNames = (ast: any) => {
+    // Add the name of the variable if it is an immutable
+    const isImmutable = ast.mutability === 'immutable'
+    const isASTNode = typeof ast.name === 'string' && typeof ast.id === 'number'
+    if (isASTNode && isImmutable) {
+      names[ast.name] = ast.id
+    }
+    // Iterate over each node
+    if (Array.isArray(ast.nodes)) {
+      for (const node of ast.nodes) {
+        findNames(node)
+      }
+    }
+    // Handle contracts that are inherited from
+    if (Array.isArray(ast.baseContracts)) {
+      for (const baseContract of ast.baseContracts) {
+        if (baseContract.baseName) {
+          const base = findContractAndSource(
+            baseContract.baseName.name,
+            buildInfo
+          )
+          findNames(base.outputSource.ast)
+        }
+      }
+    }
+  }
+
+  findNames(outputSource.ast)
+
+  let deployedBytecode = artifact.deployedBytecode
+  const presize = deployedBytecode.length
+
+  // For each of the immutables, put the value into the bytecode
+  for (const [key, value] of Object.entries(immutables)) {
```

Comment:
Could we first check that the value is `0`?

Reply:
The value can be anything - its the value of the immutable

Reply:
But what is being overwritten?

Reply:
Oh i see what you mean, check that the immutable was set to 0 in the bytecode itself

Reply:
Yeah, since the initcode never runs, that should always be the case right?